### PR TITLE
Add support for package-aware autoloading

### PR DESCRIPTION
**Pull request by aadi-stripe**
Adds package-awareness to autogen, and functionality to register packages.

### Motivation

As part of packaging work in autogen, we are requiring certain autoload definitions to be generated in the namespace of a given package.

### Test plan

See included automated tests. Tested on Stripe codebase

cc @ngroman @aisamanra

### CODE REVIEW

File: packager/packager.cc [hunk 1]
```diff
@@ -1850,51 +1860,53 @@ vector<ast::ParsedFile> Packager::run(core::GlobalState &gs, WorkerPool &workers
...
-                if (result.gotItem()) {
-                    filesProcessed++;
-                    auto &file = job.file.data(gs);
-                    core::Context ctx(gs, core::Symbols::root(), job.file);
-
-                    if (file.sourceType == core::File::Type::Normal) {
-                        job = rewritePackagedFile(ctx, move(job));
-                    } else if (file.sourceType == core::File::Type::Package) {
-                        job = rewritePackage(ctx, move(job));
+    if (!gs.runningUnderAutogen) {
```
Comment (aadi-stripe) [inline packager/packager.cc:1863]: Don't run the rest of the packager in autogen

File: packager/packager.cc [hunk 2]
```diff
@@ -1838,6 +1840,14 @@ vector<ast::ParsedFile> Packager::findPackages(core::GlobalState &gs, WorkerPool
                 } else {
                     packages.db.enterPackage(move(pkg));
                 }
+
+                if (gs.runningUnderAutogen) {
+                    it = files.erase(it);
```
Comment (aadi-stripe) [inline packager/packager.cc:1845]: Drop all package files in autogen (after packages are entered), to prevent namer errors.

**Comment by aadi-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_L4x1HxqSPJStoo
→ https://go/builds/bui_L4x1E64W3REFpf

**Comment by ngroman** [comment]
Heads up, I'll need a bit to completely work it out, but I think I have an idea for a cleaner/smaller API.

File: main/autogen/autoloader.cc [hunk 3]
```diff
@@ -222,6 +223,25 @@ string DefTree::renderAutoloadSrc(const core::GlobalState &gs, const AutoloaderC
             }
             fmt::format_to(std::back_inserter(buf), "}})\n", fullName);
         }
+
+        if (!alCfg.packagedShimPaths.empty() && gs.packageDB().isPackage(qname.nameParts)) {
```
Comment (aadi-stripe) [inline main/autogen/autoloader.cc:227]: While this code runs on every constant, `isPackage` is quite fast due to binary search and hence total overhead on Stripe codebase is not much.

File: packager/packager.cc (hunk 2 above)
Comment (ngroman) [reply packager/packager.cc:1850]: FYI erase like this is O(total len * deletes).  It's why `remove_if` exists.

Comment (aadi-stripe) [reply packager/packager.cc:1850]: Ah ok, wanted to do it in-place in the same loop, but that makes sense. Will fix.

**Review by ngroman — CHANGES_REQUESTED** [review]
What do you think about an API more like https://github.com/sorbet/sorbet/pull/5223/files

Also have some perf improvements in that implementaiton.

I think it would be cleaner and more flexible. At runtime we can make the `register_package` API do what we need.

**Comment by aadi-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_L5958qBkgC4MiY
→ https://go/builds/bui_L595Kj50ls4Rgr

**Comment by aadi-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_L59H7AA5bIB9b7
→ https://go/builds/bui_L59HLEjYWiNnRI

**Comment by aadi-stripe** [comment]
ptal @ngroman 

Made changes.

File: packager/packager.cc [hunk 4]
```diff
@@ -1849,6 +1849,12 @@ vector<ast::ParsedFile> Packager::run(core::GlobalState &gs, WorkerPool &workers
     Timer timeit(gs.tracer(), "packager");
 
     files = findPackages(gs, workers, std::move(files));
+    if (gs.runningUnderAutogen) {
+        auto it = std::remove_if(files.begin(), files.end(),
```
Comment (ngroman) [inline packager/packager.cc:1853]:
```suggestion
        // Autogen only requires package metadata. Remove the package files.
        auto it = std::remove_if(files.begin(), files.end(),
```

File: main/autogen/autoloader.cc [hunk 5]
```diff
@@ -222,6 +223,14 @@ string DefTree::renderAutoloadSrc(const core::GlobalState &gs, const AutoloaderC
             }
             fmt::format_to(std::back_inserter(buf), "}})\n", fullName);
         }
+
+        if (pkgName.exists()) {
+            ENFORCE(!gs.packageDB().empty());
+            const string_view shortName = pkgName.shortName(gs);
+            const string_view mungedName = shortName.substr(0, shortName.rfind(core::PACKAGE_SUFFIX));
```
Comment (ngroman) [inline main/autogen/autoloader.cc:230]:
```suggestion
            const string_view mungedName = shortName.substr(0, shorName.size() - core::PACKAGE_SUFFIX.size());
```
Nit: does this need a find?

**Review by ngroman — APPROVED** [review]
lgtm, see inline suggestions

File: main/autogen/autoloader.cc (hunk 5 above)
Comment (aadi-stripe) [reply main/autogen/autoloader.cc:230]: No, will work with backwards index math as well.

**Pull request merged**